tests/drivers/sensor/ina230/src/ina230_emul.c

```diff
@@ -30,23 +29,38 @@ struct ina230_emul_cfg {
 	uint16_t addr;
 };
 
+struct ina230_emul_data;
+
+typedef struct ina230_reg *(*get_register_t)(struct ina230_emul_data *, int);
+
 struct ina230_emul_data {
+	get_register_t get_register;
+
 	struct ina230_reg ina230_regs[INA230_REGISTER_COUNT];
+	struct ina230_reg ina236_regs[INA236_REGISTER_COUNT];
 };
 
-static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
+static struct ina230_reg *ina230_get_register(struct ina230_emul_data *data, int reg)
```

Comment:
I still don't see how it should work if someone requests the reg address from INA230 using the INA236 variant. `ina230_get_register` is still common for both variants as well as `ina230_regs` and `ina236_regs`. IMO. we should define INA23X_regs array only for specified variants and check what is what. I see this something like that:

<pre>
static struct ina230_reg *ina230_get_register(struct ina230_emul_data *data, int reg)
{
	for (int i = 0; i < data->regs_len; i++) {
		if (data->regs[i].id == reg) {
			return &data->regs[i];
		}
	}

	return NULL;
}
</pre>

I would add a `regs` pointer pointing to the regs array for a specific variant. This requires distinguishing between variants in the `INA230_EMUL` macro to correctly assign the regs array (you did something similar before with `is_ina236` variable).

Reply:
> I still don't see how it should work if someone requests the reg address from INA230 using the INA236 variant. `ina230_get_register` is still common for both variants as well as `ina230_regs` and `ina236_regs`. IMO. we should define INA23X_regs array only for specified variants and check what is what. I see this something like that:
> 
> static struct ina230_reg *ina230_get_register(struct ina230_emul_data *data, int reg)
> {
> 	for (int i = 0; i < data->regs_len; i++) {
> 		if (data->regs[i].id == reg) {
> 			return &data->regs[i];
> 		}
> 	}
> 
> 	return NULL;
> }
> I would add a `regs` pointer pointing to the regs array for a specific variant. This requires distinguishing between variants in the `INA230_EMUL` macro to correctly assign the regs array (you did something similar before with `is_ina236` variable).

Oh, I  misunderstood you before, you are correct about it. We can create `base_regs` and `extra_regs`, where `base_regs` will store the original INA230 registers and their default values, and the `extra_regs`  will hold the "overloaded" registers with different defaults and possibly some extra registers. This array must be **0xff** bytes long so we can hold all possible addresses. Then the `get_register` function could look like this:

```c
static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
{
        struct ina230_reg * c_reg = data->extra_regs;
        for (size_t i = 0; i <  ARRAY_SIZE(data->extra_regs); i++,c_reg++) {
                // valid entry?
                if (!c_reg->bytes) {
                        break;
                }

                if (c_reg->id == reg) {
                        return c_reg;
                }
        }

        c_reg = data->base_regs;
        for (size_t i = 0; i <  INA230_REGISTER_COUNT; i++,c_reg++) {
                if (c_reg->id == reg) {
                        return c_reg;
                }
        }

        return NULL;
}
```

The initialization can look like this:

```c
#define CREATE_INA236_REGS                                                                         \
        {                                                                                          \
                {INA236_REG_MANUFACTURER_ID, 2, 0x449}, {INA236_REG_DEVICE_ID, 2, 0xa080},         \
                        {INA230_REG_MASK, 2, 0x10},                                                \
        }

#define INA230_EMUL(n, v)                                                                          \
        static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
                .addr = DT_INST_REG_ADDR(n),                                                       \
        };                                                                                         \
        static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
                .ina230_regs =                                                                     \
                        {                                                                          \
                                {INA230_REG_CONFIG, 2, 0x4127},                                    \
                                {INA230_REG_SHUNT_VOLT, 2, 0},                                     \
                                {INA230_REG_BUS_VOLT, 2, 0},                                       \
                                {INA230_REG_POWER, 2, 0},                                          \
                                {INA230_REG_CURRENT, 2, 0},                                        \
                                {INA230_REG_CALIB, 2, 0},                                          \
                                {INA230_REG_MASK, 2, 0},                                           \
                                {INA230_REG_ALERT, 2, 0},                                          \
                        },                                                                         \
                .ina236_regs = COND_CODE_0(v, ({}), (CREATE_INA23##v##_REGS)),                     \
                .get_register = ina23##v##_get_register,                                           \
        };                                                                                         \
        EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
                            &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)
```

I am just not sure about the `COND_CODE_0` part where it is not a flag that is being checked. What do you think about this solution?

Reply:
Let's try to simplify it ;) Could you test the following pseudo-code?

<pre>
struct ina230_emul_data {
	struct ina23x_config *ina23x_config;
	const size_t ina23x_config_len;
};

static struct ina23x_config ina230_config[] = {
	{INA230_REG_CONFIG, 2, 0x4127},                                    
	{INA230_REG_SHUNT_VOLT, 2, 0},                                     
	{INA230_REG_BUS_VOLT, 2, 0},                                       
	{INA230_REG_POWER, 2, 0},                                          
	{INA230_REG_CURRENT, 2, 0},                                        
	{INA230_REG_CALIB, 2, 0},                                          
	{INA230_REG_MASK, 2, 0},                                           
	{INA230_REG_ALERT, 2, 0},                                     
};		

static struct ina23x_config ina236_config[] = {
	{INA236_REG_MANUFACTURER_ID, 2, 0x449},
	{INA236_REG_DEVICE_ID, 2, 0xa080},
};		
          			
#define INA230_EMUL(n, id, v)                                                                          \
	static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
		.addr = DT_INST_REG_ADDR(n),                                                       \
	};                                                                                         \
	static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
		.ina23x_config = &type##_config; \	
		.ina23x_config_len = ARRAY_SIZE(type##_config); \	
	};                                                                                         \
	EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
			    &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)

#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_ina230
DT_INST_FOREACH_STATUS_OKAY_VARGS(INA230_EMUL, ina230, 0)

#undef DT_DRV_COMPAT
#define DT_DRV_COMPAT ti_ina236
DT_INST_FOREACH_STATUS_OKAY_VARGS(INA230_EMUL, ina236, 6)


static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
{
	for (int i = 0; i < data->ina23x_config_len; i++) {
		if (data->ina23x_config[i]->id == reg) {
			return &data->ina23x_config[i];
		}
	}

	return NULL;
}
<pre>

Reply:
> Let's try to simplify it ;) Could you test the following pseudo-code?
> 
> struct ina230_emul_data {
> 	struct ina23x_config *ina23x_config;
> 	const size_t ina23x_config_len;
> };
> 
> static struct ina23x_config ina230_config[] = {
> 	{INA230_REG_CONFIG, 2, 0x4127},                                    
> 	{INA230_REG_SHUNT_VOLT, 2, 0},                                     
> 	{INA230_REG_BUS_VOLT, 2, 0},                                       
> 	{INA230_REG_POWER, 2, 0},                                          
> 	{INA230_REG_CURRENT, 2, 0},                                        
> 	{INA230_REG_CALIB, 2, 0},                                          
> 	{INA230_REG_MASK, 2, 0},                                           
> 	{INA230_REG_ALERT, 2, 0},                                     
> };		
> 
> static struct ina23x_config ina236_config[] = {
> 	{INA236_REG_MANUFACTURER_ID, 2, 0x449},
> 	{INA236_REG_DEVICE_ID, 2, 0xa080},
> };		
>           			
> #define INA230_EMUL(n, id, v)                                                                          \
> 	static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
> 		.addr = DT_INST_REG_ADDR(n),                                                       \
> 	};                                                                                         \
> 	static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
> 		.ina23x_config = &type##_config; \	
> 		.ina23x_config_len = ARRAY_SIZE(type##_config); \	
> 	};                                                                                         \
> 	EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
> 			    &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)
> 
> #undef DT_DRV_COMPAT
> #define DT_DRV_COMPAT ti_ina230
> DT_INST_FOREACH_STATUS_OKAY_VARGS(INA230_EMUL, ina230, 0)
> 
> #undef DT_DRV_COMPAT
> #define DT_DRV_COMPAT ti_ina236
> DT_INST_FOREACH_STATUS_OKAY_VARGS(INA230_EMUL, ina236, 6)
> 
> 
> static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
> {
> 	for (int i = 0; i < data->ina23x_config_len; i++) {
> 		if (data->ina23x_config[i]->id == reg) {
> 			return &data->ina23x_config[i];
> 		}
> 	}
> 
> 	return NULL;
> }

If I integrated your code correctly then this is not working for INA236.

How can this work with registers for INA230 when 236 is used? This code only takes a pointer for `ina236_config` registers: 

```c
static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
        .ina23x_config = &type##_config; \	
        .ina23x_config_len = ARRAY_SIZE(type##_config); \	
};
```

P.S.: is it OK to share the register values (we take the pointer to a static array) between instances of the same type of emulated devices?

Reply:
It takes the pointer to the specified reg array based on the id. Why would you want to use registers from INA230 when INA236 is being used?

Reply:
Because the tests are using them for mocks and to have a complete set of registers defined by chip. How else can we test the different resolutions of the chips?

What exactly are we trying to achieve? I thought that I needed to create a complete virtual INA236 device with all its registers so we could mock them in tests if needed.

Reply:
Oh, I missed that completely. In that case, we can use your previous approach or extend the `ina236_config` about those registers. I'm sorry for the confusion.

Reply:
It would, I just did not want to mix that up, because it would not fit the array (`INA230_REGISTER_COUNT` is 8). Also, I think `get_register` should return `NULL` when a register not provided by the device is requested.

Reply:
Are the registers split into two arrays blocking the pull? Can I do something to have this approved?

Reply:
Some of the checks failed, try to rebase with the latest main and force push. Regarding the implementation, I'm still thinking about the pre-pre version with 2 separate for loops (the current one seems too complicated to me though).

Reply:
Ok, what if we solve it like this:

```c
struct ina230_emul_data {
	struct ina230_reg *regs;
};

static struct ina230_reg *get_register(struct ina230_emul_data *data, int reg)
{
	struct ina230_reg *c_reg = data->regs;

	while (c_reg->bytes) {
		if (c_reg->id == reg) {
			return c_reg;
		}
		c_reg++;
	}

	return NULL;
}

#define CREATE_INA230_REGS                                                                         \
	{INA230_REG_CONFIG, 2, 0x4127}, {INA230_REG_SHUNT_VOLT, 2, 0},                             \
		{INA230_REG_BUS_VOLT, 2, 0}, {INA230_REG_POWER, 2, 0}, {INA230_REG_CURRENT, 2, 0}, \
		{INA230_REG_CALIB, 2, 0}, {INA230_REG_MASK, 2, 0}, {INA230_REG_ALERT, 2, 0},

#define CREATE_INA236_REGS                                                                         \
	CREATE_INA230_REGS, {INA236_REG_MANUFACTURER_ID, 2, 0x449},                                \
	{                                                                                          \
		INA236_REG_DEVICE_ID, 2, 0xa080                                                    \
	}

#define INA230_EMUL(n, v)                                                                          \
	static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
		.addr = DT_INST_REG_ADDR(n),                                                       \
	};                                                                                         \
	static struct ina230_reg ina23##v##_regs_##n[] = {CREATE_INA23##v##_REGS, {}};             \
	static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
		.regs = (struct ina230_reg *)ina23##v##_regs_##n};                                 \
	EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
			    &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)
```

With this solution there is only one array and one loop to lookup the register in the mocked device. There is a little readability issue in `CREATE_INA23x_REGS` macros, but this can be hopefully handled better later.

Reply:
Yea, readability is pretty bad, but the single array implementation looks much better now. 

btw. did you use clang formatter or is it done by hand?

Reply:
Unfortunately, it is the output from clang-format-20.

Reply:
You can try to improve that by hand as clang formatting is not a must-have.

Reply:
Is this better?

```c
/* clang-format off */

#define CREATE_INA230_REGS                                                                         \
	{INA230_REG_CONFIG, 2, 0x4127},                                                            \
	{INA230_REG_SHUNT_VOLT, 2, 0},                                                             \
	{INA230_REG_BUS_VOLT, 2, 0},                                                               \
	{INA230_REG_POWER, 2, 0},                                                                  \
	{INA230_REG_CURRENT, 2, 0},                                                                \
	{INA230_REG_CALIB, 2, 0},                                                                  \
	{INA230_REG_MASK, 2, 0},                                                                   \
	{INA230_REG_ALERT, 2, 0}

#define CREATE_INA236_REGS                                                                         \
	CREATE_INA230_REGS,                                                                        \
	{INA236_REG_MANUFACTURER_ID, 2, 0x449},                                                    \
	{INA236_REG_DEVICE_ID, 2, 0xa080}                                                          \

#define INA230_EMUL(n, v)                                                                          \
	static const struct ina230_emul_cfg ina23##v##_emul_cfg_##n = {                            \
		.addr = DT_INST_REG_ADDR(n),                                                       \
	};                                                                                         \
	static struct ina230_reg ina23##v##_regs_##n[] = {CREATE_INA23##v##_REGS, {}};             \
	static struct ina230_emul_data ina23##v##_emul_data_##n = {                                \
		.regs = (struct ina230_reg *)ina23##v##_regs_##n};                                 \
	EMUL_DT_INST_DEFINE(n, ina230_emul_init, &ina23##v##_emul_data_##n,                        \
			    &ina23##v##_emul_cfg_##n, &ina230_emul_api_i2c, NULL)

/* clang-format on */
```

Reply:
Looks pretty good now!